android/java/org/chromium/chrome/browser/playlist/PlaylistHostActivity.java

```diff
@@ -0,0 +1,465 @@
+                    if (option == PlaylistOptions.REMOVE_ALL_OFFLINE_DATA) {
+                        if (playlistOptionsModel.getAllPlaylistModels() != null) {
+                            for (PlaylistModel playlistModel :
+                                    playlistOptionsModel.getAllPlaylistModels()) {
+                                mPlaylistService.removeLocalDataForItemsInPlaylist(
+                                        playlistModel.getId());
+                            }
+                        }
+                    } else if (option == PlaylistOptions.DOWNLOAD_ALL_PLAYLISTS_FOR_OFFLINE_USE) {
+                        mPlaylistService.getAllPlaylists(playlists -> {
+                            for (Playlist playlist : playlists) {
+                                for (PlaylistItem playlistItem : playlist.items) {
+                                    mPlaylistService.recoverLocalDataForItem(playlistItem.id, true,
+                                            tempPlaylistItem
+                                            -> {
+
+                                            });
+                                }
+                            }
+                        });
+                    }
+                });
+
+        mPlaylistViewModel.getPlaylistItemOption().observe(
+                PlaylistHostActivity.this, playlistItemOption -> {
+                    if (mPlaylistService == null) {
+                        return;
+                    }
+                    PlaylistOptions option = playlistItemOption.getOptionType();
+                    if (option == PlaylistOptions.MOVE_PLAYLIST_ITEM) {
+                        showMoveOrCopyPlaylistBottomSheet();
+                    } else if (option == PlaylistOptions.COPY_PLAYLIST_ITEM) {
+                        showMoveOrCopyPlaylistBottomSheet();
+                    } else if (option == PlaylistOptions.DELETE_ITEMS_OFFLINE_DATA) {
+                        mPlaylistService.removeLocalDataForItem(
+                                playlistItemOption.getPlaylistItemModel().getId());
+                        loadPlaylist(playlistItemOption.getPlaylistId());
+                    } else if (option == PlaylistOptions.OPEN_IN_NEW_TAB) {
+                        openPlaylistInTab(
+                                false, playlistItemOption.getPlaylistItemModel().getPageSource());
+                    } else if (option == PlaylistOptions.OPEN_IN_PRIVATE_TAB) {
+                        openPlaylistInTab(
+                                true, playlistItemOption.getPlaylistItemModel().getPageSource());
+                    } else if (option == PlaylistOptions.DELETE_PLAYLIST_ITEM) {
+                        mPlaylistService.removeItemFromPlaylist(playlistItemOption.getPlaylistId(),
+                                playlistItemOption.getPlaylistItemModel().getId());
+                        loadPlaylist(playlistItemOption.getPlaylistId());
```

Comment:
Once add/delete/rename one item, here recreating whole list and re-populate views. Which is costly and will see the flicker.
There are two ways one is through
1. Update list data.add(insertIndex, item), data.remove(removeIndex);
2. add/remove particular view by using adapter.notifyItemInserted,(removeIndex);, adapter.notifyItemRemoved(removeIndex);

Other way we can use ListAdapter, DiffUtil

Reply:
@sujitacharya2005 we can do it only when we are concerned about the UI changes only. with above changes, we need to also update backend apis with the update and it should be sync. i agree with you that we should update only a part of the changes but we would improve on it as we go through the changes.